### PR TITLE
Integrate data shreds

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
We are using blobs for data plane. Big blobs with packet drops cause network to fall over. We reduced the blob size to mitigate that problem. Since blobs cannot have partial entries, it impacts overall performance. Also, if entries get bigger than blob size, it can never be transmitted.

#### Summary of Changes
Use shreds for dataplane

#5294

### CODE REVIEW

File: core/src/blocktree.rs [hunk 1]
```diff
@@ -980,6 +1134,77 @@ impl Blocktree {
...
+                shred_chunk.len(),
+                look_for_last_shred
+            );
+
+            // Break if we didn't find the last shred (as more data is required)
+            if look_for_last_shred {
+                break;
+            }
+
+            if let Ok(deshred) = Shredder::deshred(&shred_chunk) {
```
Comment (rob-solana) [inline core/src/blocktree.rs]: sure would be nice for deshred to return a cursor (implements Read) that could be passed to bincode::deserialize_from()...

File: core/src/blocktree.rs [hunk 2]
```diff
@@ -980,6 +1134,77 @@ impl Blocktree {
...
+                let mut payload_offset = 0;
+                while let Ok(entry) = bincode::deserialize(&deshred.payload[payload_offset..]) {
```
Comment (rob-solana) [inline core/src/blocktree.rs]: serialize a vector of entries into the writer, and this loop goes away

File: core/src/broadcast_stage/broadcast_utils.rs [hunk 3]
```diff
@@ -100,18 +100,17 @@ pub(super) fn entries_to_blobs(
...
-        entries.iter().for_each(|(entry, _)| {
-            trace!("Shredding entry {:#?}", entry);
-            let data = bincode::serialize(entry).unwrap();
-            trace!("Writing to shred {:#?}", &data);
-            shredder.write(&data).unwrap();
-        });
+        let data = bincode::serialize(entries).unwrap();
+        let mut offset = 0;
+        while offset < data.len() {
+            offset += shredder.write(&data[offset..]).unwrap();
```
Comment (rob-solana) [inline core/src/broadcast_stage/broadcast_utils.rs]: does shredder.write() do partial writes?

this looks like you're making a shred block per vector of entries?

Comment (pgarg66) [reply core/src/broadcast_stage/broadcast_utils.rs]: Yes, the shredder write fills up to the MTU size shred, and returns how many bytes were used. The caller has to call it again with the leftover data.

Comment (rob-solana) [reply core/src/broadcast_stage/broadcast_utils.rs]: why?

**Pull request merged**